### PR TITLE
Relocate environment variable injection to before the interpreter is run

### PR DESCRIPTION
**Pull request by erikwilliamson**
As per https://github.com/pantsbuild/pex/issues/2224

### CODE REVIEW

**Comment by jsirois** [comment]
@erikwilliamson thanks for working on this. Please add a test to prove this works. You likely can just pile in the original PR test: https://github.com/pantsbuild/pex/pull/1948/files#diff-2b86577053b9bb459d54416c397bf7d1fa598becf7a5cc9e856290fb0427881e

You can then `tox -eformat-run,typecheck` to run black + mypy and `tox -epyXY-integration -- -ktest_inject_env_and_args` to run your tests via pytest. Here XY is the Python major minor of a Python you have installed, for example `tox -epy311-integration`. You can use `./dtox` instead of tox with the rest of the command line being the same to run the tests in a docker environment that has all the Pythons Pex supports pre-installed. The 1st time you run, using `BASE_MODE=pull ./dtox` will likely speed things up by pulling the base docker image instead of building it.

**Comment by erik-at-techsanity** [comment]
@jsirois I'm sorry that this is taking me so long. I'm having trouble getting any tests to run successfully, even from the main branch. I'm totally new to Tox, too which isn't speeding things up.

I'm getting the following when running dtox.sh:

```
erik@mini pex % BASE_MODE=pull ./dtox.sh -epy311-integration
1ff12c11995a5b21b7b2774ccc974664997c955b: Pulling from pantsbuild/pex/base
... truncated the output ...
#6 0.182 + addgroup --gid=20 staff
#6 0.216 addgroup: The group `staff' already exists.

```

And while I know what this means, I'm not sure what the appropriate fix is - should I create a separate user and a different group for it to be in on my dev machine?

**Comment by jsirois** [comment]
@erikwilliamson I wanted to check in and make sure you had what you needed to run `tox -epyXY-integration -- -ktest_inject_env_and_args` (https://github.com/pantsbuild/pex/pull/2260#issuecomment-1743401346). No rush of course, but without being able to run that it might be hard to work on adding a test.

**Comment by jsirois** [comment]
I'm not a part of the Pants slack. Since you now can run tests, if you can run tox `-epyXY-integration -- -k your_new_test_name`, even if that fails for perplexing reasons, it's probably time to add the test to this PR. Once you do that I'll approve the CI workflow and we can have a shared failure experience to talk about.

**Review by jsirois — COMMENTED** [review]
@erikwilliamson this LGTM mod commented out tests. I'll let CI fly, but please circle back to get those un-commented / updated to work with this change.

**Comment by jsirois** [comment]
For the formatting shard, `tox -eformat-run,typecheck` is what you want to pass locally.

**Comment by erikwilliamson** [comment]
I'm getting this on OSX 14.1 - I can fire up a linux VM to make it go away in the meantime though

Typechecking 338 files using Python 3.11.6 against Python 3.12 ...
pex/compatibility.py: note: In function "cpu_count":
pex/compatibility.py:163:23: error: Module has no attribute
"sched_getaffinity"  [attr-defined]
                cpu_set = os.sched_getaffinity(0)

**Comment by jsirois** [comment]
@erikwilliamson if that's the only error, you should be good to go since it was only the format check that failed in CI (https://github.com/pantsbuild/pex/actions/runs/6776741887/job/18420675544?pr=2260#step:4:160) and not the type check.

... and I <3 macOS.

**Comment by kaos** [comment]
> ... and I <3 macOS.

soon I'll believe you 😂

**Comment by erikwilliamson** [comment]
> @erikwilliamson this LGTM mod commented out tests. I'll let CI fly, but please circle back to get those un-commented / updated to work with this change.

Whoops - sorry - I missed committing them earlier. Incoming...

File: tests/integration/test_inject_env_and_args.py [hunk 1]
```diff
@@ -240,8 +240,9 @@ def assert_message(
     assert_message(b"Hello, world!")
     assert_message(b"42", MESSAGE="42")
 
-    # Switching away from the built-in entrypoint should disable injected args and env.
-    assert {"args": ["foo", "bar"], "MESSAGE": None} == json.loads(
+    # Switching away from the built-in entrypoint should disable injected args but not the en env.
+    assert {"args": ["foo", "bar"], "MESSAGE": "Hello, world!"} == json.loads(
+        # assert {"MESSAGE": "Hello, world!"} == json.loads(
```
Comment (jsirois) [inline tests/integration/test_inject_env_and_args.py:245]: Kill this commented out line.

File: tests/integration/test_inject_env_and_args.py [hunk 2]
```diff
@@ -240,8 +240,9 @@ def assert_message(
     assert_message(b"Hello, world!")
     assert_message(b"42", MESSAGE="42")
 
-    # Switching away from the built-in entrypoint should disable injected args and env.
-    assert {"args": ["foo", "bar"], "MESSAGE": None} == json.loads(
+    # Switching away from the built-in entrypoint should disable injected args but not the en env.
```
Comment (jsirois) [inline tests/integration/test_inject_env_and_args.py:243]: Fix the typo at the end of the sentence. Should be: ... not the env.

**Review by jsirois — APPROVED** [review]
Thanks @erikwilliamson, this looks good with two small fixes.

File: pex/venv/installer.py [hunk 3]
```diff
@@ -829,7 +829,10 @@ def sys_executable_paths():
                 if pex_interpreter
                 else pex_overrides.get("PEX_MODULE", {entry_point!r} or PEX_INTERPRETER_ENTRYPOINT)
             )
-
+                  
+            for name, value in {inject_env!r}:
```
Comment (jsirois) [inline pex/venv/installer.py:833]: Actually, on closer review, shouldn't this move all the way up to just before line 820? It looks like `PEX_SCRIPT=<script>` / PEXes built with `-c <script>` still don't get env vars exposed in this venv case unlike the `pex.pex.PEX` case (zipapp), which you fixed properly above.

Comment (jsirois) [reply pex/venv/installer.py:833]: @erikwilliamson I'm going to get out the 2.1.151 release with `--exclude` support today before a longish break. If you want time to work on this last issue with the PR; you've got it. If you'd rather it get in this afternoon's release, I can collaborate with an edit if you wish.

Comment (erikwilliamson) [reply pex/venv/installer.py:833]: ok - I'll tackle this in about 45 minutes or so to see if we can get it done today!

**Comment by jsirois** [comment]
Alright @erikwilliamson this is now available in Pex 2.1.151:
+ https://pypi.org/project/pex/2.1.151/
+ https://github.com/pantsbuild/pex/releases/tag/v2.1.151

Thanks for your contribution.

**Comment by erikwilliamson** [comment]
Thanks so much for guiding me, @jsirois ! Your help and patience were greatly appreciated.

**Pull request merged**